rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs

```diff
@@ -0,0 +1,1089 @@
+                LastIndexedSnapshot {
+                    sequence: Some(4),
+                    at_block: 90,
+                }
+            );
+
+            // Expect the range to be:
+            // (start, tip)
+            let range = cursor.get_next_range().await.unwrap().unwrap();
+            let expected_range = 190..=200;
+            assert_eq!(range, expected_range);
+
+            // Update the cursor with no found logs.
+            cursor
+                .update(
+                    vec![
+                        // There's a gap - we missed a log at sequence 5.
+                        (MockSequencedData::new(6), log_meta_with_block(195)),
+                    ],
+                    expected_range,
+                )
+                .await
+                .unwrap();
+
+            // Expect a rewind to occur back to the last indexed snapshot's block number.
+            assert_eq!(
+                cursor.current_indexing_snapshot,
+                TargetSnapshot {
+                    sequence: 5,
+                    at_block: 90,
+                }
+            );
+            assert_eq!(
+                cursor.last_indexed_snapshot,
+                LastIndexedSnapshot {
+                    sequence: Some(4),
+                    at_block: 90,
+                }
+            );
+        }
+
+        /// Tests when the cursor is so behind the tip that it'll need to index multiple ranges, but by the time
+        /// it gets to the target snapshot, it realizes it missed a log and needs to rewind.
+        #[tracing_test::traced_test]
+        #[tokio::test]
+        async fn test_handles_unexpected_logs() {
+            // Starts with current snapshot at sequence 5, block 90
+            let mut cursor = get_cursor().await;
+
+            // Pretend like the tip is 100, and a message occurred at block 95.
+
+            cursor.latest_sequence_querier = Arc::new(MockLatestSequenceQuerier {
+                latest_sequence_count: Some(6),
+                tip: 100,
+            });
+
+            // Expect the range to be:
+            // (start, start + chunk_size)
+            let range = cursor.get_next_range().await.unwrap().unwrap();
+            let expected_range = 90..=100;
+            assert_eq!(range, expected_range);
+
+            // Update the cursor with some paritally bogus logs:
+            // - A log at sequence 4, which was already indexed and should be ignored
+            // - Two logs of sequence 5, i.e. duplicated
+            // - A log at sequence 6, which is unexpected, but tolerated nonetheless
+            cursor
+                .update(
+                    vec![
+                        (MockSequencedData::new(4), log_meta_with_block(90)),
+                        (MockSequencedData::new(5), log_meta_with_block(95)),
+                        (MockSequencedData::new(5), log_meta_with_block(95)),
+                        (MockSequencedData::new(6), log_meta_with_block(100)),
```

Comment:
If you pass the sequences out of order, e.g. `4`, `5`, `6`, `5`, the test will fail bc of the dedupe behavior I mentioned above